chitchat/src/state.rs

```diff
@@ -421,70 +510,64 @@ impl ClusterState {
         scheduled_for_deletion: &HashSet<&ChitchatId>,
     ) -> Delta {
         let mut stale_nodes = SortedStaleNodes::default();
-        let mut nodes_to_reset = Vec::new();
 
         for (chitchat_id, node_state) in &self.node_states {
             if scheduled_for_deletion.contains(chitchat_id) {
                 continue;
             }
-            let Some(node_digest) = digest.node_digests.get(chitchat_id) else {
-                stale_nodes.insert(chitchat_id, node_state);
+
+            let digest_max_version: Version = digest
+                .node_digests
+                .get(chitchat_id)
+                .map(|node_digest| node_digest.max_version)
+                .unwrap_or(0u64);
+
+            if node_state.max_version <= digest_max_version {
+                // Our version is actually older than the version of the digest.
+                // We have no update to offer.
                 continue;
-            };
-            // TODO: We have problem here. If after the delta we end up with a max version that is
```

Comment:
FYI, I believe that this problem is still present in the new version, which can lead to a node always being reset and never catching up to the max_version of the other nodes.

e.g:
Node A has last_gc=1000, many kvs in [0, 2000], and max_version=2000.
Node B was disconnected and has max_version=500.
B sends a SYN with max_version=500, A decides that it should send "should_reset" delta with last_gc=1000, from_version=0, and all the kvs in [0, 2000].
Now, if all the kvs that fit in the message have versions < 1000, B will only apply thoses deltas up to a version < 1000, and have max_version<1000. The next time it sends its digest to A, A will again decide it should send a "should_reset" delta and again send the kvs starting at 0.

The same can happen if A only has non-removed kvs in the range [0, 999].

Also, since "should_reset" deltas are sent first in the packet based on the scuttle-depth heuristic, if they reach the size of the packet, those should_reset updates sent to B will also prevent indefinitely updates to the kvs in B's cluster state for the other nodes.

So maybe preserve/extend the original comment somewhere?

Reply:
You are absolutely right. I also noticed the problem when writing down what I thought the invariant were and made an update.

In a nutshell, we do no need to reset if `digest_last_gc_version >= self.last_gc_version`.